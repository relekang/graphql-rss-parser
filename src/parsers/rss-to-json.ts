import { parser } from 'rss-to-json'
import _debug from 'debug'

import { ParserError, NotAFeedError } from '../errors'
import { Item, ParserResponse } from '../types'

const debug = _debug('graphql-rss-parser:parsers:rss-to-json')

export async function parse(feed: string): Promise<ParserResponse> {
  try {
    debug('starting to parse')
    const parsed = await parser(feed)
    debug('done parsing')

    return {
      parser: 'RSS_TO_JSON',
      title: parsed.title,
      description: parsed.description,
      link: parsed.link,
      feedLink: parsed.feedLink,
      entries: parsed.items.map((item: any): Item => {
        return Object.assign({}, item, {
          pubDate: new Date(item.created).toISOString(),
          categories: typeof item.category === 'string' ? [item.category] : item.category || [],
          guid: item.id,
        })
      }),
    }
  } catch (error) {
    debug('parsing failed with error', error)
    if (
      error.toString().includes('There are errors in your xml') ||
      error.toString().includes("Cannot read property 'item' of undefined")
    ) {
      throw new NotAFeedError()
    }
    throw new ParserError(error, 'RSS_TO_JSON')
  }
}
