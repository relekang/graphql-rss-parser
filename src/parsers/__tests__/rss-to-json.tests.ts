/* eslint-env jest */
import { parse } from '../rss-to-json'
import mockAxios from '../../__tests__/mockAxios'

test('should parse string from rolflekang.com/feed.xml', async () => {
  const fixture = await mockAxios({ url: 'https://rolflekang.com/feed.xml' })

  expect(await parse(fixture.data)).toMatchSnapshot()
})
