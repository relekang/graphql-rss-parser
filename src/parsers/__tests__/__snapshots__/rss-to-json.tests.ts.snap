// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should parse string from rolflekang.com/feed.xml 1`] = `
Object {
  "description": "undefined",
  "entries": Array [
    Object {
      "author": undefined,
      "categories": Array [
        "letsencrypt",
        "acme",
        "certbot",
        "ansible",
      ],
      "category": Array [
        "letsencrypt",
        "acme",
        "certbot",
        "ansible",
      ],
      "content": null,
      "created": 1604793600000,
      "description": undefined,
      "enclosures": Array [],
      "guid": undefined,
      "id": undefined,
      "link": "https://rolflekang.com/using-certbot-with-ansible",
      "pubDate": "2020-11-08T00:00:00.000Z",
      "published": 1604793600000,
      "title": "Using certbot with Ansible",
      "url": "https://rolflekang.com/using-certbot-with-ansible",
    },
    Object {
      "author": undefined,
      "categories": Array [
        "ansible",
        "devops",
        "ops",
        "infrastructure-as-code",
      ],
      "category": Array [
        "ansible",
        "devops",
        "ops",
        "infrastructure-as-code",
      ],
      "content": null,
      "created": 1599868800000,
      "description": undefined,
      "enclosures": Array [],
      "guid": undefined,
      "id": undefined,
      "link": "https://rolflekang.com/ansible-handlers-in-loops",
      "pubDate": "2020-09-12T00:00:00.000Z",
      "published": 1599868800000,
      "title": "Using Ansible handlers in loops",
      "url": "https://rolflekang.com/ansible-handlers-in-loops",
    },
    Object {
      "author": undefined,
      "categories": Array [
        "nextjs",
        "javascript",
        "curl",
        "nodejs",
      ],
      "category": Array [
        "nextjs",
        "javascript",
        "curl",
        "nodejs",
      ],
      "content": null,
      "created": 1590451200000,
      "description": undefined,
      "enclosures": Array [],
      "guid": undefined,
      "id": undefined,
      "link": "https://rolflekang.com/serving-plain-text-with-nextjs",
      "pubDate": "2020-05-26T00:00:00.000Z",
      "published": 1590451200000,
      "title": "Serving text/plain for curl with Next",
      "url": "https://rolflekang.com/serving-plain-text-with-nextjs",
    },
    Object {
      "author": undefined,
      "categories": Array [
        "networking",
        "unifi",
        "ubiquiti",
        "homelab",
      ],
      "category": Array [
        "networking",
        "unifi",
        "ubiquiti",
        "homelab",
      ],
      "content": null,
      "created": 1588464000000,
      "description": undefined,
      "enclosures": Array [],
      "guid": undefined,
      "id": undefined,
      "link": "https://rolflekang.com/wireless-uplinks-with-unifi",
      "pubDate": "2020-05-03T00:00:00.000Z",
      "published": 1588464000000,
      "title": "Wireless uplinks with Unifi",
      "url": "https://rolflekang.com/wireless-uplinks-with-unifi",
    },
    Object {
      "author": undefined,
      "categories": Array [
        "git",
      ],
      "category": "git",
      "content": null,
      "created": 1573689600000,
      "description": undefined,
      "enclosures": Array [],
      "guid": undefined,
      "id": undefined,
      "link": "https://rolflekang.com/using-git-commits-instead-of-stash",
      "pubDate": "2019-11-14T00:00:00.000Z",
      "published": 1573689600000,
      "title": "Using git commits instead of git stash",
      "url": "https://rolflekang.com/using-git-commits-instead-of-stash",
    },
    Object {
      "author": undefined,
      "categories": Array [
        "gatsby",
        "canvas",
        "javascript",
      ],
      "category": Array [
        "gatsby",
        "canvas",
        "javascript",
      ],
      "content": null,
      "created": 1551139200000,
      "description": undefined,
      "enclosures": Array [],
      "guid": undefined,
      "id": undefined,
      "link": "https://rolflekang.com/twitter-cards-for-gatsby-posts",
      "pubDate": "2019-02-26T00:00:00.000Z",
      "published": 1551139200000,
      "title": "Twitter cards for Gatsby posts",
      "url": "https://rolflekang.com/twitter-cards-for-gatsby-posts",
    },
    Object {
      "author": undefined,
      "categories": Array [
        "gatsby",
        "javascript",
      ],
      "category": Array [
        "gatsby",
        "javascript",
      ],
      "content": null,
      "created": 1550620800000,
      "description": undefined,
      "enclosures": Array [],
      "guid": undefined,
      "id": undefined,
      "link": "https://rolflekang.com/related-articles-with-gatsby",
      "pubDate": "2019-02-20T00:00:00.000Z",
      "published": 1550620800000,
      "title": "Related articles with Gatsby",
      "url": "https://rolflekang.com/related-articles-with-gatsby",
    },
    Object {
      "author": undefined,
      "categories": Array [
        "reasonml",
        "native",
      ],
      "category": Array [
        "reasonml",
        "native",
      ],
      "content": null,
      "created": 1549929600000,
      "description": undefined,
      "enclosures": Array [],
      "guid": undefined,
      "id": undefined,
      "link": "https://rolflekang.com/creating-a-cli-with-reason-native",
      "pubDate": "2019-02-12T00:00:00.000Z",
      "published": 1549929600000,
      "title": "Creating a CLI with Reason native",
      "url": "https://rolflekang.com/creating-a-cli-with-reason-native",
    },
    Object {
      "author": undefined,
      "categories": Array [
        "graphql",
        "testing",
        "javascript",
      ],
      "category": Array [
        "graphql",
        "testing",
        "javascript",
      ],
      "content": null,
      "created": 1495929600000,
      "description": undefined,
      "enclosures": Array [],
      "guid": undefined,
      "id": undefined,
      "link": "https://rolflekang.com/testing-simple-graphql-services",
      "pubDate": "2017-05-28T00:00:00.000Z",
      "published": 1495929600000,
      "title": "Testing simple GraphQL services",
      "url": "https://rolflekang.com/testing-simple-graphql-services",
    },
    Object {
      "author": undefined,
      "categories": Array [
        "running",
      ],
      "category": "running",
      "content": null,
      "created": 1483315200000,
      "description": undefined,
      "enclosures": Array [],
      "guid": undefined,
      "id": undefined,
      "link": "https://rolflekang.com/lets-run-some-races",
      "pubDate": "2017-01-02T00:00:00.000Z",
      "published": 1483315200000,
      "title": "Let's run some races this year",
      "url": "https://rolflekang.com/lets-run-some-races",
    },
    Object {
      "author": undefined,
      "categories": Array [
        "ipad",
        "development-environment",
      ],
      "category": Array [
        "ipad",
        "development-environment",
      ],
      "content": null,
      "created": 1483228800000,
      "description": undefined,
      "enclosures": Array [],
      "guid": undefined,
      "id": undefined,
      "link": "https://rolflekang.com/software-development-on-an-ipad",
      "pubDate": "2017-01-01T00:00:00.000Z",
      "published": 1483228800000,
      "title": "Software development on an iPad",
      "url": "https://rolflekang.com/software-development-on-an-ipad",
    },
    Object {
      "author": undefined,
      "categories": Array [
        "linter",
        "linting",
        "development",
      ],
      "category": Array [
        "linter",
        "linting",
        "development",
      ],
      "content": null,
      "created": 1465516800000,
      "description": undefined,
      "enclosures": Array [],
      "guid": undefined,
      "id": undefined,
      "link": "https://rolflekang.com/filtering-lint-errors",
      "pubDate": "2016-06-10T00:00:00.000Z",
      "published": 1465516800000,
      "title": "Filtering lint errors",
      "url": "https://rolflekang.com/filtering-lint-errors",
    },
    Object {
      "author": undefined,
      "categories": Array [
        "travel",
        "interrail",
      ],
      "category": Array [
        "travel",
        "interrail",
      ],
      "content": null,
      "created": 1437868800000,
      "description": undefined,
      "enclosures": Array [],
      "guid": undefined,
      "id": undefined,
      "link": "https://rolflekang.com/interrail-summer",
      "pubDate": "2015-07-26T00:00:00.000Z",
      "published": 1437868800000,
      "title": "Interrail summer",
      "url": "https://rolflekang.com/interrail-summer",
    },
    Object {
      "author": undefined,
      "categories": Array [
        "atom",
        "writing",
        "procrastination",
      ],
      "category": Array [
        "atom",
        "writing",
        "procrastination",
      ],
      "content": null,
      "created": 1432339200000,
      "description": undefined,
      "enclosures": Array [],
      "guid": undefined,
      "id": undefined,
      "link": "https://rolflekang.com/writing-latex-in-atom",
      "pubDate": "2015-05-23T00:00:00.000Z",
      "published": 1432339200000,
      "title": "Writing (latex) in Atom",
      "url": "https://rolflekang.com/writing-latex-in-atom",
    },
    Object {
      "author": undefined,
      "categories": Array [
        "python",
        "testing",
        "mocking",
      ],
      "category": Array [
        "python",
        "testing",
        "mocking",
      ],
      "content": null,
      "created": 1419897600000,
      "description": undefined,
      "enclosures": Array [],
      "guid": undefined,
      "id": undefined,
      "link": "https://rolflekang.com/rmoq",
      "pubDate": "2014-12-30T00:00:00.000Z",
      "published": 1419897600000,
      "title": "rmoq - a request mock cache",
      "url": "https://rolflekang.com/rmoq",
    },
    Object {
      "author": undefined,
      "categories": Array [
        "django",
        "nopassword",
        "django-nopassword",
        "authentication",
      ],
      "category": Array [
        "django",
        "nopassword",
        "django-nopassword",
        "authentication",
      ],
      "content": null,
      "created": 1412294400000,
      "description": undefined,
      "enclosures": Array [],
      "guid": undefined,
      "id": undefined,
      "link": "https://rolflekang.com/django-nopassword-one-point-o",
      "pubDate": "2014-10-03T00:00:00.000Z",
      "published": 1412294400000,
      "title": "django-nopassword reached 1.0",
      "url": "https://rolflekang.com/django-nopassword-one-point-o",
    },
    Object {
      "author": undefined,
      "categories": Array [
        "this",
        "static-site",
        "pelican",
        "jekyll",
      ],
      "category": Array [
        "this",
        "static-site",
        "pelican",
        "jekyll",
      ],
      "content": null,
      "created": 1399420800000,
      "description": undefined,
      "enclosures": Array [],
      "guid": undefined,
      "id": undefined,
      "link": "https://rolflekang.com/moving-to-pelican",
      "pubDate": "2014-05-07T00:00:00.000Z",
      "published": 1399420800000,
      "title": "Moving the blog to pelican",
      "url": "https://rolflekang.com/moving-to-pelican",
    },
    Object {
      "author": undefined,
      "categories": Array [
        "gruntjs",
        "cookiecutter",
        "tumblr",
      ],
      "category": Array [
        "gruntjs",
        "cookiecutter",
        "tumblr",
      ],
      "content": null,
      "created": 1391299200000,
      "description": undefined,
      "enclosures": Array [],
      "guid": undefined,
      "id": undefined,
      "link": "https://rolflekang.com/building-tumblr-theme-with-grunt",
      "pubDate": "2014-02-02T00:00:00.000Z",
      "published": 1391299200000,
      "title": "Building a Tumblr theme with GruntJS",
      "url": "https://rolflekang.com/building-tumblr-theme-with-grunt",
    },
    Object {
      "author": undefined,
      "categories": Array [
        "db",
        "backup",
        "redis",
      ],
      "category": Array [
        "db",
        "backup",
        "redis",
      ],
      "content": null,
      "created": 1382572800000,
      "description": undefined,
      "enclosures": Array [],
      "guid": undefined,
      "id": undefined,
      "link": "https://rolflekang.com/backup-routines-for-redisdb",
      "pubDate": "2013-10-24T00:00:00.000Z",
      "published": 1382572800000,
      "title": "Backup routine for Redis",
      "url": "https://rolflekang.com/backup-routines-for-redisdb",
    },
    Object {
      "author": undefined,
      "categories": Array [
        "django",
        "custom user",
      ],
      "category": Array [
        "django",
        "custom user",
      ],
      "content": null,
      "created": 1369180800000,
      "description": undefined,
      "enclosures": Array [],
      "guid": undefined,
      "id": undefined,
      "link": "https://rolflekang.com/django-app-custom-user",
      "pubDate": "2013-05-22T00:00:00.000Z",
      "published": 1369180800000,
      "title": "Making your Django app ready for custom users",
      "url": "https://rolflekang.com/django-app-custom-user",
    },
    Object {
      "author": undefined,
      "categories": Array [
        "pypstats",
        "pypi",
        "stats",
        "geek-tool",
      ],
      "category": Array [
        "pypstats",
        "pypi",
        "stats",
        "geek-tool",
      ],
      "content": null,
      "created": 1362268800000,
      "description": undefined,
      "enclosures": Array [],
      "guid": undefined,
      "id": undefined,
      "link": "https://rolflekang.com/stats-are-fun",
      "pubDate": "2013-03-03T00:00:00.000Z",
      "published": 1362268800000,
      "title": "Stats are fun",
      "url": "https://rolflekang.com/stats-are-fun",
    },
    Object {
      "author": undefined,
      "categories": Array [
        "alfredapp",
        "jekyll",
        "shell script",
      ],
      "category": Array [
        "alfredapp",
        "jekyll",
        "shell script",
      ],
      "content": null,
      "created": 1361836800000,
      "description": undefined,
      "enclosures": Array [],
      "guid": undefined,
      "id": undefined,
      "link": "https://rolflekang.com/alfred-scripts-for-jekyll",
      "pubDate": "2013-02-26T00:00:00.000Z",
      "published": 1361836800000,
      "title": "Alfred scripts for Jekyll",
      "url": "https://rolflekang.com/alfred-scripts-for-jekyll",
    },
    Object {
      "author": undefined,
      "categories": Array [
        "tags",
        "postgres",
        "array",
        "pgarray",
        "django",
      ],
      "category": Array [
        "tags",
        "postgres",
        "array",
        "pgarray",
        "django",
      ],
      "content": null,
      "created": 1359244800000,
      "description": undefined,
      "enclosures": Array [],
      "guid": undefined,
      "id": undefined,
      "link": "https://rolflekang.com/use-array-in-postgresql",
      "pubDate": "2013-01-27T00:00:00.000Z",
      "published": 1359244800000,
      "title": "Use array in postgresql with django",
      "url": "https://rolflekang.com/use-array-in-postgresql",
    },
    Object {
      "author": undefined,
      "categories": Array [
        "puppet",
        "web.py",
        "nginx",
        "uwsgi",
        "github",
        "github-hooks",
      ],
      "category": Array [
        "puppet",
        "web.py",
        "nginx",
        "uwsgi",
        "github",
        "github-hooks",
      ],
      "content": null,
      "created": 1357171200000,
      "description": undefined,
      "enclosures": Array [],
      "guid": undefined,
      "id": undefined,
      "link": "https://rolflekang.com/apply-puppet-configuration-automatically",
      "pubDate": "2013-01-03T00:00:00.000Z",
      "published": 1357171200000,
      "title": "Apply puppet automatically",
      "url": "https://rolflekang.com/apply-puppet-configuration-automatically",
    },
    Object {
      "author": undefined,
      "categories": Array [
        "wikipedia",
        "css",
        "readability",
        "stylebot",
      ],
      "category": Array [
        "wikipedia",
        "css",
        "readability",
        "stylebot",
      ],
      "content": null,
      "created": 1354406400000,
      "description": undefined,
      "enclosures": Array [],
      "guid": undefined,
      "id": undefined,
      "link": "https://rolflekang.com/readable-wikipedia",
      "pubDate": "2012-12-02T00:00:00.000Z",
      "published": 1354406400000,
      "title": "Readable Wikipedia",
      "url": "https://rolflekang.com/readable-wikipedia",
    },
    Object {
      "author": undefined,
      "categories": Array [
        "django",
        "no-password",
        "email-authentication",
        "authentication",
      ],
      "category": Array [
        "django",
        "no-password",
        "email-authentication",
        "authentication",
      ],
      "content": null,
      "created": 1348790400000,
      "description": undefined,
      "enclosures": Array [],
      "guid": undefined,
      "id": undefined,
      "link": "https://rolflekang.com/django-nopassword",
      "pubDate": "2012-09-28T00:00:00.000Z",
      "published": 1348790400000,
      "title": "django-nopassword",
      "url": "https://rolflekang.com/django-nopassword",
    },
    Object {
      "author": undefined,
      "categories": Array [
        "arduino",
        "github",
        "git",
        "web.py",
      ],
      "category": Array [
        "arduino",
        "github",
        "git",
        "web.py",
      ],
      "content": null,
      "created": 1347753600000,
      "description": undefined,
      "enclosures": Array [],
      "guid": undefined,
      "id": undefined,
      "link": "https://rolflekang.com/the-github-lamp",
      "pubDate": "2012-09-16T00:00:00.000Z",
      "published": 1347753600000,
      "title": "The Github lamp",
      "url": "https://rolflekang.com/the-github-lamp",
    },
  ],
  "feedLink": undefined,
  "link": "https://rolflekang.com",
  "parser": "RSS_TO_JSON",
  "title": "Writing by Rolf Erik Lekang",
}
`;
